### DISCUSSION TITLE
Detect and Generate WP_CACHE_KEY_SALT Automatically

### DISCUSSION THREAD

**Discussion by jordantrizz** [rhubarbgroup/.github#19] [Ideas (ROC)]
This feature request is two-fold and can be considered a safeguard for when WordPress instances are copied and utilized on the same server without specifying a WP_CACHE_KEY_SALT. 

Granted, this is a rare scenario but should be considered as some non-technical people may end up running multiple instances of WordPress on the same server without defining WP_CACHE_KEY_SALT. Resulting in multiple instances using the same cache keys.

Perhaps the WP_CACHE_KEY_SALT if not defined should be based off get_site_url() or something else that is unique. For instance Litespeed Object Cache using the md5 of the full path to it's object.lib.php file. Which will ultimately be unique if on the same server. Unless of course you're running a immutable multi-site sharing the same plugin files. But you could also take 3-5 different strings to create a unique string that is then md5'd.

Here's the code.

https://github.com/litespeedtech/lscache_wp/blob/d1a0aef92424d1c32886f4d594e094ade141d6ad/src/object.lib.php#LL427C1-L437C75

**Comment by tillkruss** [comment]
@jordantrizz We have considered this in the past, the issue is:

1. What if the `get_site_url()` changes?
2. What happens to existing sites without a prefix, they would all have double the data until they flush the cache?

**Reply by jordantrizz** [reply]
> 1. What if the `get_site_url()` changes?

Then the prefix should change, or am I misunderstanding? Here's what Litespeed did.

```
	public function __construct() {
		$this->_object_cache = \LiteSpeed\Object_Cache::cls();

		$this->multisite = is_multisite();
		$this->blog_prefix = $this->multisite ? get_current_blog_id() . ':' : '';

		/**
		 * Fix multiple instance using same oc issue
		 * @since  1.8.2
		 */
		! defined( 'LSOC_PREFIX' ) && define( 'LSOC_PREFIX', substr( md5( __FILE__ ), -5 ) );
	}
```

They used the __FILE__ global variable which is the full path to the current file, md5'd the string and only took the last 5 chars. It is working out for them on 4 million installs. I only suggested the get_site_url() to add to the string or potentially more items that would uniquely identify the WordPress instance.

> 2. What happens to existing sites without a prefix, they would all have double the data until they flush the cache?

Do you mean if you deploy this method within the plugin today? The prefix changes to the auto-generated prefix and the data that wasn't prefixed would lay dormant until flushed manually or expire if they have a TTL? That would be a concern. I haven't dug deep to see how your plugin sets TTL's on data.

I'm sure there is a way to confirm the current state.
If - $WP_CACHE_KEY_SALT is set, do nothing
else - WP_CACHE_KEY_PREFIX isn't set, then set based on the md5 of __FILE__.

For existing installs, when the plugin is updated flush the cache, and follow the logic above. Granted if someone was not to update the plugin through WordPress and update the plugin files either by overwriting or removing and then adding. That would be problematic, you may have to detect the version currently installed and clear the cache first then let the auto prefix run.

This may not be easy to implement, and I'm not a WordPress plugin expert. I only know enough to be dangerous. The goal here would be to eliminate human error, that's all. Sorry I've been so vocal here, hopefully it hasn't added more to your plate :)

**Reply by tillkruss** [reply]
If a site doesn't have a `MaxTTL` set, basing it on the site-url or a the file's MD5 has means the Redis database would double in size. We could use the SQL database name + prefix, that would likely be unique.

> It is working out for them on 4 million installs.

Their Redis implementation is rather bad, I don't want to copy them.

**Reply by jordantrizz** [reply]
> If a site doesn't have a MaxTTL set, basing it on the site-url or a the file's MD5 has means the Redis database would double in size. We could use the SQL database name + prefix, that would likely be unique.

That would work too, most likely there wouldn't be the same database name on the same server. Unless someone is using two different remote MySQL servers with the same database name. Which would be highly unlikely, unless they were using a cluster with reads and write servers. Again, still would be the same instance in that case.

> Their Redis implementation is rather bad, I don't want to copy them.

I was only using the single function in their plugin as a reference to the idea I was suggesting. What you've suggested is a better idea versus the md5 of the file path.

**Comment by jordantrizz** [comment]
Hey Till, saw this closed. I was just wondering if you came to a solution?

**Reply by tillkruss** [reply]
The solution is to set `WP_REDIS_PREFIX` for each installation.

It's too complex to implement an automatic prefix IMO. You're welcome to submit a PR.
